platforms/core-execution/worker-processes/src/main/java/org/gradle/process/internal/worker/problem/WorkerProblemSerializer.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal.worker.problem;
+
+import org.gradle.api.problems.internal.DefaultProblem;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.DefaultSerializerRegistry;
+import org.gradle.internal.serialize.Encoder;
+import org.gradle.internal.serialize.Serializer;
+import org.gradle.internal.serialize.SerializerRegistry;
+
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+
+/**
+ * Simple serializer π {@link DefaultProblem} used by the worker/daemon communication protocol.
+ */
+public class WorkerProblemSerializer {
+    public static SerializerRegistry create() {
+        DefaultSerializerRegistry registry = new DefaultSerializerRegistry(false);
+
+        registry.register(DefaultProblem.class, new ProblemSerializer());
+        return registry;
+    }
+
+    private static class ProblemSerializer implements Serializer<DefaultProblem> {
+
+        @Override
+        public void write(Encoder encoder, DefaultProblem problem) throws Exception {
+            ObjectOutputStream objectOutputStream = new ObjectOutputStream(encoder.getOutputStream());
```

Comment:
Does it make sense to avoid Java serialization here so we don't lean _more_ into it? 

Everything I hear about Java Serialization points to it being on the way out so I feel that if we can avoid adding more code that relies on it the better.

Reply:
The issue here is that we need to serialize a complex objects (i.e. location types), and anything other than object serialization involves a lot more work, complexity, and potential pitfalls.
Also, I assume here a very low chance of classpath collision, so I felt confident just using `ObjectOutputStream`.

> everything I hear about Java Serialization points to it being on the way out 
I'm interested in where you hear this, as it seems appropriate to me: if you need an IPC channel, and you know you are running on the same intersection of the classpath being used in the objects, I think this is by far the fastest route.

Reply:
My understanding is that Java serialization is the de-facto way for worker-daemon communication. This is used extensively in test workers as well (see https://github.com/gradle/gradle/blob/master/platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/worker/ForkingTestClassProcessor.java#L120). If we want to change that, I think we should do it consistently and with a dedicated effort.

Reply:
The concept of java serialization is nice but there are some pitfalls:
https://openjdk.org/projects/amber/design-notes/towards-better-serialization

Brian Goetz is not much of a fan and that gives me concerns about the future of the feature. 

One primary problem I have with java serialiation is that there doesn't seem to be any real guarantees from the JVM that the serialization format is stable across Java versions. We are currently using this heavily for worker serialization and it seems to work but IDK I just don't like the idea of it

In terms of changes here, as long as this is used extensively elsewhere then I don't really have a big problem with it.

Reply:
Thanks for the link. That was an excellent read!